exporter/clickhouseexporter/internal/metrics_model.go

```diff
@@ -0,0 +1,158 @@
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+var supportMetricsType = [...]string{createGaugeTableSQL, createSumTableSQL, createHistogramTableSQL, createExpHistogramTableSQL, createSummaryTableSQL}
+
+type MetricsModel interface {
+	Add(metrics interface{}, name string, description string, unit string)
+	injectMetaData(metaData *MetricsMetaData)
+	insert(ctx context.Context, tx *sql.Tx, logger *zap.Logger) error
+}
+
+type MetricsMetaData struct {
+	ResAttr    map[string]string
+	ResURL     string
+	ScopeURL   string
+	ScopeInstr pcommon.InstrumentationScope
+}
+
+func CreateMetricsTable(tableName string, ttlDays uint, db *sql.DB) error {
+	var ttlExpr string
+	if ttlDays > 0 {
+		ttlExpr = fmt.Sprintf(`TTL toDateTime(TimeUnix) + toIntervalDay(%d)`, ttlDays)
+	}
+	for _, table := range supportMetricsType {
+		query := fmt.Sprintf(table, tableName, ttlExpr)
+		if _, err := db.Exec(query); err != nil {
+			return fmt.Errorf("exec create metrics table sql: %w", err)
+		}
+	}
+	return nil
+}
+
+func CreateMetricsModel(tableName string) map[pmetric.MetricType]MetricsModel {
+	metricsMap := make(map[pmetric.MetricType]MetricsModel)
+	metricsMap[pmetric.MetricTypeGauge] = &GaugeMetrics{
+		insertSQL: fmt.Sprintf(strings.ReplaceAll(insertGaugeTableSQL, "'", "`"), tableName),
+	}
+	metricsMap[pmetric.MetricTypeSum] = &SumMetrics{
+		insertSQL: fmt.Sprintf(strings.ReplaceAll(insertSumTableSQL, "'", "`"), tableName),
+	}
+	metricsMap[pmetric.MetricTypeHistogram] = &HistogramMetrics{
+		insertSQL: fmt.Sprintf(strings.ReplaceAll(insertHistogramTableSQL, "'", "`"), tableName),
+	}
+	metricsMap[pmetric.MetricTypeExponentialHistogram] = &ExpHistogramMetrics{
+		insertSQL: fmt.Sprintf(strings.ReplaceAll(insertExpHistogramTableSQL, "'", "`"), tableName),
+	}
+	metricsMap[pmetric.MetricTypeSummary] = &summaryMetrics{
+		insertSQL: fmt.Sprintf(strings.ReplaceAll(insertSummaryTableSQL, "'", "`"), tableName),
+	}
+	return metricsMap
+}
+
+func InjectMetaData(metricsMap map[pmetric.MetricType]MetricsModel, metaData *MetricsMetaData) {
+	for _, metrics := range metricsMap {
+		metrics.injectMetaData(metaData)
+	}
+}
+
+func InsertMetrics(ctx context.Context, tx *sql.Tx, metricsMap map[pmetric.MetricType]MetricsModel, logger *zap.Logger) error {
+	var errs []error
+	for _, metrics := range metricsMap {
+		errs = append(errs, metrics.insert(ctx, tx, logger))
+	}
+	return multierr.Combine(errs...)
+}
+
+func convertExemplars(exemplars pmetric.ExemplarSlice) (clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet, clickhouse.ArraySet) {
+	var (
+		attrs       clickhouse.ArraySet
+		times       clickhouse.ArraySet
+		floatValues clickhouse.ArraySet
+		intValues   clickhouse.ArraySet
+		traceIDs    clickhouse.ArraySet
+		spanIDs     clickhouse.ArraySet
+	)
+	for i := 0; i < exemplars.Len(); i++ {
+		exemplar := exemplars.At(i)
+		if !exemplar.TraceID().IsEmpty() && !exemplar.SpanID().IsEmpty() {
```

Comment:
Is the traceID really necessary to check here? Are examplars only valid if these are present? Please add a test.

Reply:
done